feeds/wifi-trunk/hostapd/patches/901-hapd-ubus-channel-switch.patch

```diff
@@ -0,0 +1,188 @@
+@@ -696,6 +698,44 @@ static int hostapd_sessions(struct ubus_
+ 	return 0;
+ }
+ 
++static int hostapd_get_chan_switch_events(struct ubus_context *ctx,
++			struct ubus_object *obj,
++		        struct ubus_request_data *req,
++			const char *method,
++                        struct blob_attr *msg)
++{
++	void *a = NULL;
++	void *t = NULL;
++
++
++	struct hostapd_chan_event_list *entry;
++	blob_buf_init(&b_ev, 0);
++	a = blobmsg_open_table(&b_ev, "chan_switch_event");
++	list_for_each_entry(entry, &chan_events->list, list)
++	{
++		t = blobmsg_open_table(&b_ev, "event");
++		blobmsg_add_u32(&b_ev, "radio_name", entry->records.band);
++		blobmsg_add_u8(&b_ev, "reason", entry->records.reason);
++		blobmsg_add_u64(&b_ev, "timestamp", entry->records.ts.tv_sec * (uint64_t)1000000);
++		blobmsg_add_u32(&b_ev, "frequency", entry->records.freq);
++		blobmsg_close_table(&b_ev, t);
++	}
++
++	blobmsg_close_table(&b_ev, a);
++	ubus_send_reply(ctx, req, b_ev.head);
++	
++	/*delete list*/
++
++	if (!list_empty(&chan_events->list)) {
++	list_for_each_entry(entry, &chan_events->list, list)
++		{
++			list_del(&entry->list);
++		}
++	}
++
++	return 0;
++}
++
+ static int
+ hostapd_switch_chan(struct ubus_context *ctx, struct ubus_object *obj,
+ 		    struct ubus_request_data *req, const char *method,
+@@ -731,6 +771,9 @@ hostapd_switch_chan(struct ubus_context
+ 
+ 	if (hostapd_switch_channel(hapd, &css) != 0)
+ 		return UBUS_STATUS_NOT_SUPPORTED;
++
++	hostapd_ubus_handle_channel_switch_event(hapd->iface,HOSTAPD_UBUS_HIGH_INTERFERENCE,  css.freq_params.freq);
++
+ 	return UBUS_STATUS_OK;
+ #undef SET_CSA_SETTING
+ }
+@@ -1320,6 +1363,7 @@ static const struct ubus_method bss_meth
+ #ifdef NEED_AP_MLME
+ 	UBUS_METHOD("switch_chan", hostapd_switch_chan, csa_policy),
+ 	UBUS_METHOD_NOARG("get_sessions", hostapd_sessions),
++	UBUS_METHOD_NOARG("get_chan_switch_events", hostapd_get_chan_switch_events),
+ 	UBUS_METHOD_NOARG("clear_sessions", hostapd_clear_sessions),
+ 	UBUS_METHOD("clear_session", hostapd_clear_session, client_session_del_policy),
+ #endif
+@@ -1361,6 +1405,7 @@ void hostapd_ubus_add_bss(struct hostapd
+ 	if (asprintf(&name, "hostapd.%s", hapd->conf->iface) < 0)
+ 		return;
+ 
++	INIT_LIST_HEAD(&chan_events->list);
+ 	avl_init(&hapd->ubus.rt_events, avl_compare_sess_id, false, NULL);
+         avl_init(&hapd->ubus.banned, avl_compare_macaddr, false, NULL);
+ 	obj->name = name;
+@@ -1489,6 +1534,28 @@ ubus_event_cb(struct ubus_notify_request
+ 	ureq->resp = ret;
+ }
+ 
++
++void hostapd_ubus_handle_channel_switch_event(struct hostapd_iface *iface, int reason,
++			       int freq)
++{
++	struct hostapd_chan_event_list *rec = NULL;
++	struct timespec ts;
++	clock_gettime(CLOCK_REALTIME, &ts);
++	rec = os_zalloc(sizeof(struct hostapd_chan_event_list));
++
++	rec->records.reason = reason;
++	rec->records.band = iface->freq;
++	rec->records.ts = ts;
++	rec->records.freq = freq;
++
++	//if (list_empty(&chan_events->list)) {
```

Comment:
dead code, how often do i need to say this ..

Reply:
fixed